Show Building Grades In UI

vkoves opened the pull request:
# Description

A final PR to show grades, with all the fixes to make it final:

1. **Fixed Up Grades** - Colton spotted some weirdness with grades, and I fixed how grades handled consistent reporting, and not using graded weighting
2. **Show Grades On The Homepage & Building Details** - implemented the designs for grades, including report cards, adding grades to each sub-section, and some minor design tweaks (like a thicker border on the tiles)
3. **Added Grade Explainer Page** - added a new page that explains how and why we grade buildings.
4. **Hid Report Card For Anomalous Buildings** - for buildings with grading anomalies, like [1830 N Winchester](https://electrifychicago.net/building/1830-n-winchester-ave/), the report card is blurred behind a dialog

<details>
  <summary> <strong>Anomalous Building Example</strong> </summary>
  
![sample-flagged-building](https://github.com/user-attachments/assets/9e5e6ee7-a0a9-491b-81c7-201f07b92773)
</details>

| Before | After |
| --- | --- |
| ![Screenshot from 2025-03-05 01-11-15](https://github.com/user-attachments/assets/e0adb369-d167-4e91-b506-4924287eda57) | ![Screenshot from 2025-03-05 01-11-18](https://github.com/user-attachments/assets/ca89fc70-3983-4293-a9d3-eb04a03c3baf) |
| ![Screenshot from 2025-03-05 01-13-17](https://github.com/user-attachments/assets/8f50b0db-da89-41fd-8099-aaeae0bb672b) | ![Screenshot from 2025-03-05 01-13-14](https://github.com/user-attachments/assets/c0bdad23-1b5f-41e5-bc90-591ec367c224) |

**See designs in [Figma](https://www.figma.com/design/rJQiz1waySWCRN9NdGtjtg/Electrify-Chicago?node-id=207-2&t=gjucCyKJMbflpgPB-0)**

Fixes #129

# Testing Instructions

Spot checked specific buildings to confirm grades seem sound, and tested in desktop and mobile to confirm the report card anchors work properly.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

<!-- PR template modified from: https://embeddedartistry.com/blog/2017/08/04/a-github-pull-request-template-for-your-projects/ -->

vkoves commented:
@colton-lapp - you can see Sasha's grade notebook here: `src/data/research/building_grade_eda.ipynb`

vkoves commented:
@colton-lapp - Bertha and I played a bit more with the quartile distribution, and it definitely seems better. But we also saw some odd things, like buildings with the existing systems that got all As but still got a B 🤔 

![Screenshot from 2025-02-25 21-54-39](https://github.com/user-attachments/assets/c844ed6e-2522-48ea-b887-1a14e94986f7)

I'd love any of your thoughts and analysis!

colton-lapp commented:
I tinkered around in the notebook you linked and made two graphs that show what I understand are the criteria we care about with grading:

![3dgraph](https://github.com/user-attachments/assets/999d022e-97fb-4e61-8157-adb64b8f4580)
![3 subplots](https://github.com/user-attachments/assets/9dfc0a6e-276c-49dd-9054-47b938a4d9e6)

I haven't dug into the formula for how these overall grades get computed, but it does seem like maybe we are favoring energy mix too much. It seems like the only A's are all electric and the only F's don't use any electricity. 

![as and fs](https://github.com/user-attachments/assets/f658ea8a-873a-428c-8984-1b53177d81ff)

colton-lapp commented:
> I tinkered around in the notebook you linked and made two graphs that show what I understand are the criteria we care about with grading:

Looking at the 3D graph I uploaded and toggling on different groups, I think it could make sense to loosen the importance of energy mix. There are a lot of buildings with 0 missing reports, <10 GHGIntensity and 50%+ electricity mix that should probably be an A? For the Ds, there are many buildings with <50% electric mix, 3+ missing reports, and GHG Intensity >10 which could probably be Fs. 

Maybe we could use these graphs to hardcode some rules? That could make the grades concrete and give avenues for improvement and we could show these graphs as justification. Maybe a weighted formula like:

Total points = constant + electricity mix - GHG Intensity - 10 x # missed reports 
And then assign letter cutoffs for each grade that give us  nice distributions. 

It feels a little hand wavey, but it would ultimately be transparent to people, rooted in fundamentals of climate friendly buildings, and we could show correlations of the points and the energy start score to prove that it has some support.

vkoves commented:
@colton-lapp - digging into the grading more, I'm torn if energy mix is over-weighted, but I wonder if the weighting isn't working properly. I see a few things:

1. This line makes the average grade, but it just averages everything - it doesn't weight the missing records to 10%, for example, and I think that also means the energy mix is counting for 33% rather than 40%:

```py

    # Overall numerical grade, average of all percentile grades:
    graded_df["AvgPercentileGrade"] = graded_df[
        [
            "GHGIntensityPercentileGrade",
            "EnergyMixWeightedPctSumPercentileGrade",
            "MissingRecordsCountPercentileGrade"
        ]
    ].mean(axis=1)
```

32 The missing records is still using a relative percentile (better than X% of buildings), even though my thought was it's just a linear grade (there is a perfect value here, after all). The records letter grade is not impacted;

![image](https://github.com/user-attachments/assets/330bd6e0-f70e-4e33-99cd-a19ec58df1a9)

That means that it appears merch mart's overall score is actually getting dinged for 100% reporting, somehow 🤔 

I think a good approach might be to write some unit tests for certain buildings and what we expect each sub-score to be. I'm pretty hesitant to hard code a more specific formula, as I think the overall weighted percentages idea is sound, but we need to make sure it's working correctly.

I think the more  I dig into actual examples though, I feel like the sub-scores generally work as intended. With the Monadnock, for example.

- It gets a D on energy mix because it has a higher % of gas use than average (it's energy mix percentile score is 33)
- It gets an A on reporting
- It gets an A on GHG intensity

However, looking at the numbers, it got a 44 for the reporting score, and I think that is what moved it from a B to a C:

![Screenshot from 2025-02-27 23-46-16](https://github.com/user-attachments/assets/0865e139-497f-49a6-a129-2ecddd83b492)

If I score it myself, it would be:

```
97 (GHG Intensity Percentile) × 0.5 + 33 (Energy Mix Percentile) × 0.4 + 100 (Perfect Reporting) ×0.1 = 71.7
```

And 71.7 is a B, not a C. I bet the missing records count is artificially deflating a bunch of scores, but the energy mix is still a very strong factor, and prevents gas heated buildings (which basically have to have > 50% gas share) from getting an A. In my mind though, that is probably correct, because even a very efficient gas building can have huge emissions (esp. if it's large, like merch mart) and will require costly retrofits to make it net zero carbon, even on a clean grid.

vkoves commented:
@colton-lapp - I fixed the reporting score being wrong (perfect scores were marked as poor) and the weighted average not being used. I think this fixed an overall grade deflation, and now the base grades line up with our old quintiles, but with some improvements. Marina Towers moved from B -> A, Merch Mart from B -> A, and Monadnock from C -> B. I couldn't get the notebook running, but by using the search (which I added grade filters to), I can see there are now:

- 227 A Buildings
- 913 B Buildings
- 1,547 C Buildings
- 648 D Buildings
- 47 F Buildings

In my mind, this actually seems pretty sound. Instead of an even distribution, you have a few exceptional buildings (227 / 3382 or 6.7% A), many good buildings  (913 / 3382 or 26.9% B), mostly C buildings (45%) with some D (19%) and very few horrible F buildings. Since both GHG intensity and energy mix are percentiles compared to other buildings, to be an A building you have to be _well above average_ in both, and to be an F building you have to be _well below average_ in both. In my mind, this makes sense, since a building that has the average % gas use (~60% based on Art Institute getting a C for that) but is very efficient, it's still a highly polluting building. 

It's also worth noting that **emissions intensity also skews towards energy mix** - this is because the emissions intensity of electricity is lower than gas. This might mean those two variables are a bit redundant, and we could instead look at EUI and Energy Mix, but most folks aren't familiar with EUI, so I'm worried that would be more technical and introduce another variable.

vkoves commented:
Alright, we manually QA-d this, so merging this in!
